### ISSUE TITLE
TPT count query performance

### ISSUE THREAD

**Issue by samcarton** [dotnet/efcore#31304]
A question regarding the count query generated when a TPT entity count is performed.  

Assume the structure as per the docs, but with some other FK on the derived entity:
```
public class Blog
{
    public int BlogId { get; set; }
    public string Url { get; set; }
}

public class RssBlog : Blog
{
    public string RssUrl { get; set; }
    public int SiteId { get; set; } // FK to another type, we will filter on this
}

```

When I perform a simple filtered count on RssBlog:

```
var siteRssBlogCount = await myContext.RssBlogs
                .AsNoTracking()
                .Where(x => x.SiteId == someSiteId)
                .CountAsync();
```

it generates a query that looks like this:

```
SELECT count ( * ) :: int 
FROM Blogs b 
INNER JOIN RssBlogs r ON b.Id = r.Id 
WHERE r.SiteId = <site id param>
```

### The question:

Is there any reason it couldn't bypass the base class table `Blogs` and just do a straight count on `RssBlogs`? The FK on the derived table ensures that a base table row must exist for each derived table row.
eg:

```
SELECT count ( * ) :: int 
FROM RssBlogs  r 
WHERE r.SiteId = <site id param>
```

We have real-world cases where this join to the base type table causes query performance to be significantly slower. See explain output summaries:

```
-- For EF generated count query that joins to base table:
-- QUERY PLAN
-- Planning Time: 0.480 ms
-- Execution Time: 241.689 ms

-- For a count query that only uses the derived type table:
-- QUERY PLAN
-- Planning Time: 0.083 ms
-- Execution Time: 14.113 ms

```

I understand it's probably not straightforward to _just_ implement a change like this, but I was curious if there were any cases that I wasn't aware of that would prevent this from working, or if there was a nice way to tell EF to **not** use the base table in this case.

### Provider and version information

EF Core version: 7.0.9
Database provider: Npgsql
Target framework: .NET 7.0

**Comment by roji** [comment]
Yeah, that does seems like a possible optimization - since you're not actually projecting blogs out and aren't accessing any properties on the base type, it should be possible to prune out Blogs entirely. As you suspect this wouldn't be trivial to implement, but it certainly makes sense.

We generally discourage TPT as an inheritance mapping strategy for requiring many joins, which adversely affect performance; although in this specific case the join could be avoided, in many others it cannot. So if you care enough about performance to want this join removed, I'd advise considering another strategy (either TPH or TPC) as an alternative.
